### ISSUE TITLE
[Spree Upgrade] Refactor all uses of spree_products.count_on_hand

### ISSUE THREAD

**Issue by oeoeaio** [openfoodfoundation/openfoodnetwork#2013]
Several instances of dead code that use these attributes have been identified as per: #2004 

We need to work out an approach for finding and dealing with other uses of this method.

Likely to be related to #2014

**Comment by sauloperez** [comment]
`spree_products.count_on_hand` is just the aggregation of the `count_on_hand` of the variants of the product:

https://github.com/openfoodfoundation/spree/blob/86bf87f1b1e1b299edc8cd10a2486e44ba0a3987/core/app/models/spree/product.rb#L253-L260

**Comment by sauloperez** [comment]
I just spotted a use of `spree_products.count_on_hand` in:

https://github.com/openfoodfoundation/openfoodnetwork/blob/46d2d6cb8ac9425485cc9635cc5dbc515583151f/app/views/spree/admin/overview/_enterprises_producers_tab.html.haml#L28-L35

That method is defined as follows:
https://github.com/openfoodfoundation/openfoodnetwork/blob/46d2d6cb8ac9425485cc9635cc5dbc515583151f/app/models/enterprise.rb#L206-L208

I wonder if that color change is worth the pain since the method is only used there.

**Comment by sauloperez** [comment]
We overrode the dashboard's controller `Spree::Admin::OverviewController` by means of a `class_eval` in [app/controllers/spree/admin/overview_controller_decorator.rb](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/controllers/spree/admin/overview_controller_decorator.rb). There we have custom logic to either render `single_enterprise_dashboard` or `multi_enterprise_dashboard`.

**Comment by sauloperez** [comment]
Given all the problems aforementioned I would simply remove that tab. It's something easy to do but I'd like to check somehow (using analytics?) if someone uses it.

**Comment by myriamboure** [comment]
Hum... @sauloperez I'm wondering one thing, for a producer who only manages their products catalog (as it is used by hubs on OFN) but doesn't manage any hub, that would mean they wouldn't have access to their enterprise from dashboard, am I right? That would be an issue I guess... isn't it possible to just have one tab "my enterprises" and put them  all in if the point is to remove the producer tab? Or is the problem to bring info on active products / products in OC? Cause in that case we can also make the dashboard simpler, I'm not sure those info are so relevant also as you don't know which products are in which OC so for me it brings more confusion than useful info.

**Comment by sauloperez** [comment]
I've found all the uses of `spree_products.count_on_hand` in https://github.com/openfoodfoundation/openfoodnetwork/pull/2183. Check the commit for the details.

**Comment by daniellemoorhead** [comment]
@sstead any thoughts on the above ☝️ ?

**Comment by sauloperez** [comment]
@myriamboure the problem is just with that Producers tab on the right. The problem comes with those counts of active products, which currently depend on a database column that in Spree 2.0 no longer exists. 

I feel like just a very small percentage of our users noticed that tab so I would consider removing it rather than reimplementing it, considering the fact that it has a negative impact on performance. Again, we don't have the data to claim no one uses it but I think we can do an educated guess and simplify the product.

**Comment by myriamboure** [comment]
I have no objection to that, we can see then if people should and reimplement something else simpler later on if there is an obvious need...
Just one clarification @sauloperez : if the "producer" tab disappear, will the "hub" tab also disappeared? I guess it should, that would make no sense having a sub table with one tab. So then we would only have "my enterprises" with all my enterprises in it, hubs and producers mixed. Am I right?
@sstead any objection?

**Comment by sstead** [comment]
I agree, that tab would be rarely used. I never use it, so i’m comfortable to see it go.

**Comment by sauloperez** [comment]
> So then we would only have "my enterprises" with all my enterprises in it, hubs and producers mixed. Am I right?

Exactly. Just a list of enterprises.

**Comment by sauloperez** [comment]
Ok. Moving forward with this one.

**Comment by mkllnk** [comment]
https://github.com/openfoodfoundation/openfoodnetwork/pull/2183 identified all uses of `spree_products.count_on_hand` at the time. Here is an updated summary:

- `app/models/enterprise.rb`
  - https://github.com/openfoodfoundation/openfoodnetwork/blob/eb9064f5be3433ed72f708c32bccbaa635a2988d/app/models/enterprise.rb#L134
  - https://github.com/openfoodfoundation/openfoodnetwork/blob/eb9064f5be3433ed72f708c32bccbaa635a2988d/app/models/enterprise.rb#L203
- ~`app/views/spree/admin/overview/_enterprises_producers_tab.html.haml`~
- `spec/controllers/cart_controller_spec.rb`
  - https://github.com/openfoodfoundation/openfoodnetwork/blob/eb9064f5be3433ed72f708c32bccbaa635a2988d/spec/controllers/cart_controller_spec.rb#L11
- `spec/models/enterprise_spec.rb`
  - https://github.com/openfoodfoundation/openfoodnetwork/blob/eb9064f5be3433ed72f708c32bccbaa635a2988d/spec/models/enterprise_spec.rb#L345
  - https://github.com/openfoodfoundation/openfoodnetwork/blob/eb9064f5be3433ed72f708c32bccbaa635a2988d/spec/models/enterprise_spec.rb#L540
  - https://github.com/openfoodfoundation/openfoodnetwork/blob/eb9064f5be3433ed72f708c32bccbaa635a2988d/spec/models/enterprise_spec.rb#L544
  - https://github.com/openfoodfoundation/openfoodnetwork/blob/eb9064f5be3433ed72f708c32bccbaa635a2988d/spec/models/enterprise_spec.rb#L549

**Comment by mkllnk** [comment]
I just discovered more recent code using `count_on_hand`:

- https://github.com/openfoodfoundation/openfoodnetwork/blob/eb9064f5be3433ed72f708c32bccbaa635a2988d/app/models/product_import/entry_processor.rb#L78-L83
- https://github.com/openfoodfoundation/openfoodnetwork/blob/eb9064f5be3433ed72f708c32bccbaa635a2988d/app/models/product_import/entry_processor.rb#L196
- https://github.com/openfoodfoundation/openfoodnetwork/blob/eb9064f5be3433ed72f708c32bccbaa635a2988d/app/models/product_import/entry_validator.rb#L269

These require some effort to abstract. The current code operates directly on the database which is the most efficient. If we want to keep it that way, we have to change the SQL queries to work with Spree's new database schema.

**Comment by mkllnk** [comment]
I have been looking more into the Product Import code and there are a few things where I don't understand the design decisions. It is work in progress and I understand that it needs refactoring. Right now, a Spree upgrade would break it. Before I continue working on this, I would like some input from @Matt-Yorkley about the roadmap for Product Import and how we should tackle it.

**Comment by luisramos0** [comment]
Hey guys, I found this issue on the code review column in zenhub. It isn't ready for review is it?
I only found @mkllnk 's commit related to active distributors.

**Comment by mkllnk** [comment]
Thanks @luisramos0. I think this got moved around with a connected PR. It should go back to dev-ready as there is still some work to do, especially around product import.

**Comment by mkllnk** [comment]
I just found that the `ProductSerializer` is using `on_demand`. We need to check if that's needed or replace the implementation.

**Comment by sauloperez** [comment]
The only other reference I found (quick check) other than product import and migrations is https://github.com/openfoodfoundation/openfoodnetwork/blob/master/spec/javascripts/unit/bulk_product_update_spec.js.coffee#L186. I tried locally and removing the line still makes the test pass what I'd be great to understand what's the test actually doing.

Anyway, I'd say we can create an issue for that and close this Epic as soon as #2688 is closed. As you said, this is core only.
